cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -569,24 +574,25 @@ public void removeApplication(final ApplicationId appId) throws Exception {
    * @throws CannotBeDeletedException : the application cannot be deleted because of running programs
    */
   private void ensureNoRunningPrograms(final ApplicationId appId) throws CannotBeDeletedException {
-    //Check if all are stopped.
-    Iterable<ProgramRuntimeService.RuntimeInfo> runtimeInfos =
-      Iterables.filter(runtimeService.listAll(ProgramType.values()),
-                       new com.google.common.base.Predicate<ProgramRuntimeService.RuntimeInfo>() {
-                         @Override
-                         public boolean apply(ProgramRuntimeService.RuntimeInfo runtimeInfo) {
-                           return runtimeInfo.getProgramId().getParent().equals(appId);
-                         }
-                       });
-
-    Set<String> runningPrograms = new HashSet<>();
-    for (ProgramRuntimeService.RuntimeInfo runtimeInfo : runtimeInfos) {
-      runningPrograms.add(runtimeInfo.getProgramId().getProgram());
-    }
-
-    if (!runningPrograms.isEmpty()) {
+    Map<ProgramRunId, RunRecordMeta> runningPrograms = store.getRuns(ProgramRunStatus.RUNNING, null);
```

Comment:
Can't pass in the predicate, as it is based entirely on the keys, but the predicate would be based entirely on the value.

I could be wrong, but I don't think passing in the applicationId would work because all of the keys are based on a `RECORD_TYPE` and the `programId`, so passing in the applicationId would only make the filtering happen later, after the key is built from a `RECORD_TYPE` and `null` programId.

However, what do you think about this procedure?
1. Use the runtimeInfo to get anything that is running. If there is nothing running, then it is safe to delete.
2. If there are programs in the runtimeInfo, gather them into a `Set<ProgramRunId>`, and call  `store.getRuns(Set<ProgramRunId> programRunIds)`, and then check again if there are any run records that are STARTING/RUNNING.
  a. If there are, then the program is actually still running.
  b. If there aren't, then the runtimeinfo hasn't been updated, but that's okay, let the application be deleted.